src/irr/asset/CGLIWriter.cpp

```diff
@@ -0,0 +1,327 @@
+		{
+			return *(reinterpret_cast<const aType*>(data));
+		}
+
+		bool CGLIWriter::writeAsset(io::IWriteFile* _file, const SAssetWriteParams& _params, IAssetWriterOverride* _override)
+		{
+			if (!_override)
+				getDefaultOverride(_override);
+
+			SAssetWriteContext ctx{ _params, _file };
+
+			const asset::ICPUImageView* imageView = IAsset::castDown<ICPUImageView>(_params.rootAsset);
+
+			io::IWriteFile* file = _override->getOutputFile(_file, ctx, { imageView, 0u });
+
+			if (!file)
+				return false;
+
+			return writeGLIFile(file, imageView);
+		}
+
+		bool CGLIWriter::writeGLIFile(io::IWriteFile* file, const asset::ICPUImageView* imageView)
+		{
+			os::Printer::log("WRITING GLI: writing the file", file->getFileName().c_str(), ELL_INFORMATION);
+
+			const auto& imageViewInfo = imageView->getCreationParameters();
+			const auto& imageInfo = imageViewInfo.image->getCreationParameters();
+			const auto& image = imageViewInfo.image;
+
+			if (image->getRegions().length() == 0)
+			{
+				os::Printer::log("WRITING GLI: there is a lack of regions!", file->getFileName().c_str(), ELL_INFORMATION);
+				return false;
+			}
+			if (isBlockCompressionFormat(imageInfo.format))
+			{
+				os::Printer::log("WRITING GLI: the writer doesn't support any compression format!", file->getFileName().c_str(), ELL_INFORMATION);
+				return false;
+			}
+				
+
+			const bool facesFlag = doesItHaveFaces(imageViewInfo.viewType);
+			const bool layersFlag = doesItHaveLayers(imageViewInfo.viewType);
+
+			const bool floatingPointFlag = isFloatingPointFormat(imageInfo.format);
+			const bool integerFlag = isIntegerFormat(imageInfo.format);
+			const bool signedTypeFlag = isSignedFormat(imageInfo.format);
+		
+			const auto texelBlockByteSize = asset::getTexelOrBlockBytesize(imageInfo.format);
+			const auto channelsAmount = getFormatChannelCount(imageInfo.format);
+			const auto singleChannelByteSize = texelBlockByteSize / channelsAmount;
+			const auto data = reinterpret_cast<const uint8_t*>(image->getBuffer()->getPointer());
+
+			auto getTarget = [&]()
+			{
+				switch (imageViewInfo.viewType)
+				{
+					case ICPUImageView::ET_1D: return gli::TARGET_1D;
+					case ICPUImageView::ET_1D_ARRAY: return gli::TARGET_1D_ARRAY;
+					case ICPUImageView::ET_2D: return gli::TARGET_2D;
+					case ICPUImageView::ET_2D_ARRAY: return gli::TARGET_2D_ARRAY;
+					case ICPUImageView::ET_3D: return gli::TARGET_3D;
+					case ICPUImageView::ET_CUBE_MAP: return gli::TARGET_CUBE;
+					case ICPUImageView::ET_CUBE_MAP_ARRAY: return gli::TARGET_CUBE_ARRAY;
+					default: assert(0);
+				}
+			};
+
+			auto getFacesAndLayersAmount = [&]()
+			{
+				size_t layers, faces;
+				const auto arrayLayers = imageInfo.arrayLayers;
+
+				if (layersFlag)
+					layers = ((arrayLayers - 1) / 6) + 1;
+				else
+					layers = 1;
+				if (facesFlag)
+					faces = ((arrayLayers - 1) % 6) + 1;
+				else
+					faces = 1;
+
+				return std::make_pair(layers, faces);
+			};
```

Comment:
this is all wrong, you won't be able to handle (1d,2d) array textures

CUBEMAP behaves as-if its a 6 layer 2D array
CUBEMAP_ARRAY behaves as-if its a 6*N layer 2D array